### ISSUE TITLE
Support DID metadata queries with inequality operators:

### ISSUE THREAD

**Issue by astroclark** [rucio/rucio#3750]
Motivation
----------
This is a request for support for did metadata queries using inequalities like:
```
rucio list-dids-by-meta --scope blah --key>some_value
```
where `key` is something like a `gps-start-time` with a numerical value (e.g. 1276523729) associated with the DID.

The particular use case here is for LIGO/Virgo where our rucio-managed files contain time-series data and are associated with some time segment.  

Modification
------------
Introduce inequalities and integer/float values to the metadata query in [did.py](https://github.com/rucio/rucio/blob/master/lib/rucio/core/did.py#L1477)

**Comment by bari12** [comment]
Hi James,

We are currently re-writing the metadata component which should come with the next feature release. See #3096 
I think such a change should not be too complicated, but it should be addressed abstractly in the new metadata interface, so it is possible to use that as well in future metadata plugins. So it's important to get the architecture/abstraction right and future-proof.

@gabrielefronze is this something you might be interested to work on?

**Comment by gabrielefronze** [comment]
Hi,
Yes, with pleasure!

**Comment by bari12** [comment]
👍 great. Let's have a chat about this sometime after `1.23` is out. 😄

**Comment by gabrielefronze** [comment]
Sounds good to me

**Comment by gabrielefronze** [comment]
Hey there,

A little bump up of this thread, since I started working on this.
As discussed with @bari12 the road map is something in the lines of:
* figure out a common language to declare the filters from the CLI
* make `rucio list-did-by-meta` understand such language via the newly introduced https://github.com/rucio/rucio/blob/fe6eb2e9cd7e1bc9464b955aa0a7d974c961fe82/lib/rucio/core/did_meta_plugins/did_meta_plugin_interface.py#L6
* allow metadata specification at the schema level, e.g. https://github.com/rucio/rucio/blob/be2ac1cb86f3fe426b2e8aa5dd87491bce771d3c/lib/rucio/common/schema/atlas.py#L406

At the moment I am working on the first item (defining the language) here: https://github.com/gabrielefronze/rucio-inequality-engine

Basically this allows already something like:

```
rucio list-dids-by-meta --scope blah --filter "some_value < key < some_other_value"
```

This work will be then introduced in Rucio here: https://github.com/rucio/rucio/blob/fe6eb2e9cd7e1bc9464b955aa0a7d974c961fe82/lib/rucio/core/did_meta_plugins/did_column_meta.py#L227

I'll keep progress posted! 😁

**Comment by robbarnsley** [comment]
@bari12 et al.

I've been working on adding support for the JSON plugin to the engine. This isn't particularly trivial, and am a little concerned about how robust this will ultimately be.

As an example, consider two DIDs with a key `test4` assigned using the current set-metadata CLI command:

```bash
[root@ubuntu rucio]# rucio get-metadata mock:dsn_9087c7367a234a1999e2c8fa5bfebb7e --plugin json | grep test4:
test4:  test
[root@ubuntu rucio]# rucio get-metadata mock:dsn_f3268122266d44ffb9496a04f22dcce4 --plugin json | grep test4:
test4:  4.0
```

where the values are stored in the meta json column as strings.

Now, say I want to formulate a query against this json key for, e.g. "test4>3.5". I have to cast. 

The only information I have about what to cast to is in the expression value (3.5), so i'm attempting to typecast this and using the result to infer the type for the key in the database. That is to say, in this case, I will be performing a cast to a float i.e. `did_meta['test4'].as_float()`, which brings me to an issue: what is the expected behaviour if there are incompatible mixed types for the same key (as in the example above)? Currently the cast to float will correctly fail the query for  `mock:dsn_9087c7367a234a1999e2c8fa5bfebb7e` with a dbapi error (`invalid_text_representation` error for psql). 

The query will also fail if the user gives an expression with a value that is incompatible when trying to cast the key, but that is probably reasonable.

**Comment by bari12** [comment]
In general I think we need to assume that the VO enforces some kind of schema. Of course that's easy to say, perhaps we need to enable them to enforce that via Rucio. 

But I think mixed values like this should simply not be possible.

From an implementation point of view, I would just cast it based on the value you see. (You only have to cast it on the CLI anyway, via the python clients/REST you could assume that people have to define it themselves)
